src/apps/beanstalk/ethereum/beanstalk.silo.contract-position-fetcher.ts

```diff
@@ -0,0 +1,92 @@
+import { Inject } from '@nestjs/common';
+import { BigNumber, BigNumberish, ethers } from 'ethers';
+
+import { APP_TOOLKIT, IAppToolkit } from '~app-toolkit/app-toolkit.interface';
+import { PositionTemplate } from '~app-toolkit/decorators/position-template.decorator';
+import { gqlFetch } from '~app-toolkit/helpers/the-graph.helper';
+import { DefaultDataProps } from '~position/display.interface';
+import { MetaType } from '~position/position.interface';
+import { ContractPositionTemplatePositionFetcher } from '~position/template/contract-position.template.position-fetcher';
+import {
+  GetDefinitionsParams,
+  DefaultContractPositionDefinition,
+  GetTokenDefinitionsParams,
+  UnderlyingTokenDefinition,
+  GetDisplayPropsParams,
+  GetTokenBalancesParams,
+} from '~position/template/contract-position.template.types';
+
+import { BEANSTALK_ADDRESS, Token, GRAPH_URL, silos } from '../common/beanstalk.helper';
+import { Beanstalk, BeanstalkContractFactory } from '../contracts';
+import { GET_DEPOSIT } from '../graphql/getSiloDeposit';
+
+export type SiloDefinition = {
+  address: string;
+  underlying: Token[];
+  siloName: string;
+};
+
+@PositionTemplate()
+export class EthereumBeanstalkSiloContractPositionFetcher extends ContractPositionTemplatePositionFetcher<Beanstalk> {
+  groupLabel = 'Silo Deposits';
+
+  constructor(
+    @Inject(APP_TOOLKIT) protected readonly appToolkit: IAppToolkit,
+    @Inject(BeanstalkContractFactory) protected readonly beanstalkContractFactory: BeanstalkContractFactory,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract(_address: string): Beanstalk {
+    return this.beanstalkContractFactory.beanstalk({ address: BEANSTALK_ADDRESS, network: this.network });
```

Comment:
I'm quite confused why the contract is pointing towards this address when getDefinitions is using silos ?

Reply:
Maybe I don't fully understand how this is _supposed_ to work, but this was my approach.

All of Beanstalk is deployed to just one single contract (it's using the Diamond Facet pattern), so there's only one address. People can stake BEANS or BEANS:3CRV LP. Staking happen at the same one address. Staking earns you BEANS, and the earned beans are cumulative, not broken down per each item staked.

My understanding of how this contract-position fetcher works is this:
- `getDefinitions()` gives you a list definitions. A definition can represent a position that is staked, such as LP tokens staked in Curve gauge or in Convex, or a Silo in Beanstalk
- `getTokenDefinitions()` is called for each item returned by `getDefinitions()`. It should return all the tokens associated with that Definition, for ex, the underlying tokens that are being staked (for ex in an LP token), and any other earned tokens
- `getTokenBalancesPerPosition()` for each Definition (aka position, aka staked item),  get the amount of that token that is deposited. 

Note: I might use some terms interchangeably, maybe in error. To me, a definition represents a position, which effectively means something that is staked/deposited.

So if my understanding above is relatively correct, then this is why I made the following choices:

- `getContract()` returns the main, one and only, Beanstalk contract. This is because there is no other contract. This is also why I memoized it previously. I am aware that the intent of this method is to return the contract of a specific position's contract, such a distinct pool in Curve. But this doesn't work for beanstalk, so I'm always just returning the same contract.

- using `0xearnedbeans` as the address of the token returned from `getDefinitions()`. First, the earnings are cumulative across ALL silos, there's no way to get the earnings from just a single position; this is why it's a separate contract-position fetcher.
Second, to make this work, I effectively pretend there's an LP token with address `0xearnedbeans`, and this position returns just one, claimable, token.

Hope this helps